clippy_lints/src/undocumented_unsafe_blocks.rs

```diff
@@ -171,13 +136,85 @@ fn block_has_safety_comment(cx: &LateContext<'_>, block: &Block<'_>) -> bool {
     // won't work. This is to avoid dealing with where such a comment should be place relative to
     // attributes and doc comments.
 
+    span_from_macro_expansion_has_safety_comment(cx, block.span) || span_in_body_has_safety_comment(cx, block.span)
+}
+
+/// Checks if the lines immediately preceding the item contain a safety comment.
+fn item_has_safety_comment(cx: &LateContext<'_>, item: &hir::Item<'_>) -> bool {
+    if span_from_macro_expansion_has_safety_comment(cx, item.span) || span_in_body_has_safety_comment(cx, item.span) {
+        return true;
+    }
+
+    if item.span.ctxt() == SyntaxContext::root() {
+        if let Some(parent_node) = get_parent_node(cx.tcx, item.hir_id()) {
+            let mut span_before_item = None;
+            let mut hi = false;
+            if let Node::Item(parent_item) = parent_node {
+                if let ItemKind::Mod(parent_mod) = &parent_item.kind {
+                    for (idx, item_id) in parent_mod.item_ids.iter().enumerate() {
+                        if *item_id == item.item_id() {
+                            if idx == 0 {
+                                // mod A { /* comment */ unsafe impl T {} ... }
+                                // ^------------------------------------------^ gets this span
+                                // ^---------------------^ finally checks the text in this range
+                                hi = false;
+                                span_before_item = Some(parent_item.span);
+                            } else {
+                                let prev_item = cx.tcx.hir().item(parent_mod.item_ids[idx - 1]);
+                                // some_item /* comment */ unsafe impl T {}
+                                // ^-------^ gets this span
+                                //         ^---------------^ finally checks the text in this range
+                                hi = true;
+                                span_before_item = Some(prev_item.span);
+                            }
+                            break;
+                        }
+                    }
+                }
+            }
+            let span_before_item = span_before_item.unwrap();
+
+            let source_map = cx.sess().source_map();
+            if let Some(item_span) = walk_span_to_context(item.span, SyntaxContext::root())
```

Comment:
`walk_span_to_context` isn't needed here. `item.span` is already at the root context.